### PR TITLE
Correctly handle errors during `Domain.spawn`

### LINKED ISSUE #12269: possible memory-leak on failed domain backup thread installation
From #12263 there's a possible rare memory leak in `runtime/domain.c` if `install_backup_thread` fails to spawn the thread.  

@kayceesrk suggests that instead of `install_backup_thread` raising a Failure exception directly, it should return the failure and the callers of `install_backup_thread` (of which there are two, since domain 0 is a special case) need to call the domain termination functions (which will clean up the allocated resources and also raise the Failure exception).

### PR DESCRIPTION
**Pull request by gadmm**
Fix #12269 and similar issues.

We generalize the use of `sync_check_error` and make it into a `caml_check_error_exn` that allows for clean-up before raising.

This is based on top of #12407 and #12409.

### CODE REVIEW

**Comment by gadmm** [comment]
cc @kayceesrk

**Comment by gadmm** [comment]
Let turn this into a draft given that its dependency #12407 has stalled.

**Comment by gasche** [comment]
@fabbing or @OlivierNicole, is there a small chance that one of you two would be interested in reviewing this?

**Comment by gasche** [comment]
Hmm. Actually I wrote a bit too fast: I hadn't realized that #12407 was a dependency here. I don't have the details in working memory but I remember that we don't have a design consensus for #12407, so it is unlikely that we can move on without solving that first *as long as* #12407 remains a dependency. I would thus consider the present PR as "not ready for review" for now, and @fabbing and @OlivierNicole are off the hook.

@gadmm do we really need #12407 for this? I eyeballed the first fix commit quickly, and it seems to not depend on the new API.

**Comment by gadmm** [comment]
Thanks for the ongoing work around this series of PRs.

I have rebased and reviewed the present PR again, which led to more work to avoid exceptions in critical paths, then the debugging of a deadlock and consequently another rabbit hole around the interaction between systhreads and domains.

I plan to eventually split this PR in more manageable PRs, so it remains a draft at the moment. I confirm that #12407 is a dependency, so I will wait that the latter is merged to make progress. #12407 is now in a consensual form so I do not expect any aspect to slow down its review now.

Lastly, this PR makes heavy use of the `_exn` convention. Should #13013 be merged before this PR, I would greatly appreciate any assistance in adapting to the new convention as my time will be limited.

**Comment by gadmm** [comment]
I have split #13063 from this PR. The recommended course of action is:
1. Merge #12407 & #13063 (independently)
2. Let me rebase and split this PR
3. Merge this PR
4. Merge #13013 & the splitted PR (independently)

**Comment by gadmm** [comment]
#13063 has been merged, so the next blocking step is #12407

**Comment by gadmm** [comment]
I have rebased this PR and split unrelated parts into #13227. I have fixed a couple of CI issues. It is now ready for review.

During rebasing, I had to revert a couple of lines of needless changes that occurred since the last rebase. Despite this the rebasing was still a bit more complicated than expected. This PR still uses the old encoded exceptions and I propose that the conversion to the new exception monad is done in a separate PR.

**Comment by gasche** [comment]
I would prefer to see `caml_result` in this PR -- we decided to deprecate encoded exceptions to get rid of them, it feels weird to start introducing more of them in trunk, even if only as an intermediary state. I may try to help to do this myself, but I won't have time to do it in the next few days.

File: runtime/domain.c [hunk 1]
```diff
@@ -1146,114 +1147,148 @@ CAMLexport _Atomic caml_timing_hook caml_domain_terminated_hook =
...
-    (caml_result_is_exception(result) ?
-     1 /* Error */ :
-     0 /* Ok */),
-    result.data);
+  if (Is_exception_result(res_or_exn)) {
+    res = Extract_exception(res_or_exn);
+    /* [Error res] */
+    res = caml_alloc_1(1, res);
+  } else {
+    /* [Ok res_of_exn] */
```
Comment (ghost) [inline runtime/domain.c]: Did you mean
```suggestion
    /* [Ok res_or_exn] */
```
?

File: runtime/domain.c [hunk 2]
```diff
@@ -1276,13 +1318,16 @@ CAMLprim value caml_domain_spawn(value callback, value term_sync)
   init_domain_ml_values(p.ml_values, callback, term_sync);
 
   err = pthread_create(&th, 0, domain_thread_func, (void*)&p);
-
   if (err) {
-    caml_failwith("failed to create domain thread");
+    free_domain_ml_values(p.ml_values);
+    caml_check_error(err, "failed to create domain thread : pthread_create");
```
Comment (ghost) [inline runtime/domain.c]:
```suggestion
    caml_check_error(err, "failed to create domain thread: pthread_create");
```

File: runtime/sync_posix.h [hunk 3]
```diff
@@ -118,23 +118,21 @@ Caml_inline int sync_condvar_wait(sync_condvar c, sync_mutex m)
...
-
-  if (retcode == 0) return;
-  if (retcode == ENOMEM) caml_raise_out_of_memory();
-  err = caml_strerror(retcode, buf, sizeof(buf));
-  msglen = strlen(msg);
-  errlen = strlen(err);
-  str = caml_alloc_string(msglen + 2 + errlen);
+  char * err = caml_strerror(retcode, buf, sizeof(buf));
+  int msglen = strlen(msg);
+  int errlen = strlen(err);
```
Comment (ghost) [inline runtime/sync_posix.h]: While you're here,
```suggestion
  size_t msglen = strlen(msg);
  size_t errlen = strlen(err);
```

**Review by MisterDA — COMMENTED** [review]
Why not completely remove `sync_check_error_exn` and move `caml_check_error` and `caml_check_error_exn` to `runtime/fail.c`?

**Comment by gadmm** [comment]
> Why not completely remove `sync_check_error_exn` and move `caml_check_error` and `caml_check_error_exn` to `runtime/fail.c`?

I decided to preserve the posix compatibility layer in case the win32 backend is revived. But let me know if the plans are different.

File: runtime/domain.c (hunk 1 above)
Comment (gadmm) [reply runtime/domain.c]: Thanks! (that's actually a preexisting typo in temporarily reverted code)

File: runtime/sync_posix.h (hunk 3 above)
Comment (gadmm) [reply runtime/sync_posix.h]: Thanks!

**Comment by gadmm** [comment]
This is now rebased and ready to be merged in my opinion.

@gasche I welcome any help for converting this code to use `caml_result`. I have a different appreciation of what is important and I am looking forward to moving on to something else. If we need to wait, the help is also welcome for rebasing this PR on top of more recent pending PRs if necessary. Let me know if you (or anyone else) prefer pushing directly to this PR, submitting a PR to this PR on my repository, or opening a separate PR than can be merged right after this one.

**Comment by gasche** [comment]
I understand that it's my duty to work on adapting this PR to `caml_result`, and will try to work on it on my next hacking spree.

File: otherlibs/systhreads/st_stubs.c [hunk 4]
```diff
@@ -666,6 +689,7 @@ static st_retcode create_tick_thread(void)
 CAMLprim value caml_thread_new(value clos)
 {
   CAMLparam1(clos);
+  CAMLlocal1(res);
```
Comment (gasche) [inline otherlibs/systhreads/st_stubs.c]: It is difficult for me to tell why you decided to root this variable, and I find it isky to root a variable that is, at various points within the function, not a valid OCaml value. Why are we doing this? (Can we have a comment that explains it, if it is subtle?)

File: otherlibs/systhreads/st_stubs.c [hunk 5]
```diff
@@ -847,11 +880,14 @@ static struct custom_operations caml_threadstatus_ops = {
   custom_fixed_length_default
 };
 
-static value caml_threadstatus_new (void)
+static value caml_threadstatus_new_exn(void)
 {
   st_event ts = NULL;           /* suppress warning */
   value wrapper;
-  caml_check_error(st_event_create(&ts), "Thread.create");
+  value ret = st_event_create(&ts);
```
Comment (gasche) [inline otherlibs/systhreads/st_stubs.c]: This should be `int ret`, right?

File: runtime/domain.c [hunk 6]
```diff
@@ -1212,7 +1213,7 @@ static void* domain_thread_func(void* v)
   /* Cannot access p below here. */
 
   if (domain_self) {
-    install_backup_thread(domain_self);
+    install_backup_thread_exn(domain_self);
```
Comment (gasche) [inline runtime/domain.c]: I'm reading commit-by-commit and this looks it introduces an issue where failures in `install_backup_thread_exn` are dropped silently.

File: runtime/domain.c [hunk 7]
```diff
@@ -1267,6 +1268,13 @@ CAMLprim value caml_domain_spawn(value callback, value term_sync)
   if (caml_debugger_in_use)
     caml_fatal_error("ocamldebug does not support spawning multiple domains");
 #endif
+
+  /* When domain 0 first spawns a domain, the backup thread is not
+     active, we ensure it is started here. We do this before the first
+     domain is spawned in the event it should fail. */
+  value res = install_backup_thread_exn(domain_self);
+  if (Is_exception_result(res)) caml_raise(Extract_exception(res));
```
Comment (gasche) [inline runtime/domain.c]: On first read of this comment I thought that you were talking about the backup thread for the domain that is about to be created -- and it seemed odd to me to create it from the parent.  I wonder if the comment should be rephrased slightly to make it clear that we are talking about the backup thread of domain 0?

File: runtime/domain.c [hunk 8]
```diff
@@ -1147,114 +1146,147 @@ CAMLexport _Atomic caml_timing_hook caml_domain_terminated_hook =
 
 static void domain_terminate(void);
 
-static value make_finished(caml_result result)
+static value make_finished(value res_or_exn)
```
Comment (gasche) [inline runtime/domain.c]: Wait, this change is moving code from the new/better API to the old/worse API. This is the wrong direction, right?

**Review by gasche — COMMENTED** [review]
I reviewed all the preparation commits before the big change (excluded). I find reviewing this fairly exhausting, these are delicate failure paths in resource-handling C code, so I'm not sure I have the energy to review the rest of the PR in one go -- and there are things that don't look very convincing to me in the central commit.

Would you maybe be willing to split the preparatory work in its own PR, I can do the caml_result stuff there if you don't want to, and then rebase the hard bit as its own PR on top of it to be reviewed separately?

File: runtime/sync.c [hunk 9]
```diff
@@ -29,6 +29,11 @@
 /* System-dependent part */
 #include "sync_posix.h"
 
+CAMLexport void caml_check_error(int retcode, char * msg)
+{
+  return sync_check_error(retcode, msg);
+}
```
Comment (MisterDA) [inline runtime/sync.c]: This should be (even though the function is rewritten in later commits)
```c
CAMLexport void caml_check_error(int retcode, char const * msg)
{
  sync_check_error(retcode, msg);
}
```

to prevent this MSVC warning:
```
runtime/sync.c(38): warning C4098: 'caml_check_error': 'void' function returning a value
```

and the `msg` parameter should be `const`.

File: runtime/caml/fail.h [hunk 10]
```diff
@@ -73,6 +73,9 @@ struct caml_exception_context {
 
 int caml_is_special_exception(value exn);
 
+/* from runtime/sync.c */
+CAMLextern void caml_check_error(int err, char * msg);
```
Comment (MisterDA) [inline runtime/caml/fail.h]: The `msg` parameter should be `char const * msg`.

**Comment by gadmm** [comment]
Thank you for the feedback, I will come back to this when back from holidays.

**Comment by OlivierNicole** [comment]
@gadmm Gentle ping, have been able to look at this PR again?

**Comment by gadmm** [comment]
Sorry, I have not yet found the time to go back to this. I will do it eventually, but if anyone want to look at it before this, feel free and let me know.

**Comment by gadmm** [comment]
> @gadmm Gentle ping, have been able to look at this PR again?

Some news: I have found the time to go back to this PR, several times now since November 2024.

With the little time on my hands I first rebased #13227 to not depend on the present PR that seemed to take time (because #13227 fixed deadlocks and it needed cautious reviewing).

Later, yielding to the [pressure](https://github.com/ocaml/ocaml/pull/12410#discussion_r1684914426), I tried to convert the present PR to using the new fancy (and needless for runtime code) `caml_result` but this was starting to take me several occasions and sent me into a rabbit hole of making `caml_result` actually nice.

This is my last attempt to salvage this PR and to do so I made the choice of not using `caml_result` but using the perfectly fine "encoded" exceptions (though the removal of `caml_raise_if_exception` made it a bit less nice).

File: otherlibs/systhreads/st_stubs.c (hunk 4 above)
Comment (gadmm) [reply otherlibs/systhreads/st_stubs.c]: Thanks, I have rephrased it by using two different variable names so that the root contains valid values at all times (of course, with your new result type this problem would disappear). I do not think it is incorrect to not root the variable, but there is an advantage in not having to reason that the call to `thread_remove_and_free` does not call the GC.

File: otherlibs/systhreads/st_stubs.c (hunk 5 above)
Comment (gadmm) [reply otherlibs/systhreads/st_stubs.c]: Fixed, thanks

File: runtime/domain.c (hunk 6 above)
Comment (gadmm) [reply runtime/domain.c]: Indeed, and it prepares a later commit that properly deals with this exception. The intermediate commit does not behave worse than the earlier code, by the way.

File: runtime/domain.c (hunk 7 above)
Comment (gadmm) [reply runtime/domain.c]: I just clarified this comment. Also note that you commented on a documentation comment from someone else.

As a heads up, an assignment of "domain_state->tid" has been added in the meanwhile at this precise location without comment as to why it is important to do it here or is related to backup thread spawning, I am not responsible for updating this comment and would not know how to clarify it.

File: runtime/domain.c (hunk 8 above)
Comment (gadmm) [reply runtime/domain.c]: Reverting your change was the simplest thing to do to deal with the impedance mismatch between the two kinds of results, but I have now done it differently. (The adaptation to using `caml_result` everywhere will just be slightly less trivial for you than simply reverting this change to `make_finished` again.)

File: runtime/sync.c (hunk 9 above)
Comment (gadmm) [reply runtime/sync.c]: Thanks, this should be fixed.

File: runtime/caml/fail.h (hunk 10 above)
Comment (gadmm) [reply runtime/caml/fail.h]: Thanks, this has been fixed

**Comment by gadmm** [comment]
I have split and reworked the main commit for better readability. Hopefully it is much easier to understand from the diff alone now.

I do not think it makes much sense to split the PR since the first commits only make sense in preparation of the later ones.

I removed the commit "Add blocking section when waiting" because it was originally in #13227 and it might be better reviewed separately.

File: runtime/domain.c [hunk 11]
```diff
@@ -1222,6 +1222,27 @@ static void sync_result(value term_sync, value res)
   CAMLreturn0;
 }
 
+static void terminate_domain(struct domain_ml_values *ml_values,
```
Comment (gasche) [inline runtime/domain.c]: Nitpick: I think this could be called `sync_and_terminate`. (One advantage of tweaking the name slightly is that I would be less likely to confuse it with `caml_domain_terminate`.)

File: runtime/domain.c [hunk 12]
```diff
@@ -1266,22 +1287,8 @@ static void* domain_thread_func(void* v)
...
-       joining domains. The unlock is done after [caml_domain_terminate] to
-       ensure that this domain has released all of its runtime state.
-       We call [caml_mutex_unlock] directly instead of
-       [caml_ml_mutex_unlock] because the domain no longer exists at
-       this point. */
-    caml_mutex_unlock(mut);
-
+    caml_result res = caml_callback_res(unrooted_callback, Val_unit);
+    terminate_domain(ml_values, res);
     /* [ml_values] must be freed after unlocking [mut]. This ensures
```
Comment (gasche) [inline runtime/domain.c]: The refactoring in this commit makes the comment `[ml_values] must be freed after unlocking [mut]` hard to follow, because `mut` is not defined anymore locally, it is part of the `sync_and_terminate` call.

(It is not obvious to me how to rephrase it, I would let you do what you think is best. Now the code organization makes the order obvious, which is good, but this remains a very tricky failure condition whose documentation may remain valuable.)

File: runtime/domain.c [hunk 13]
```diff
@@ -2209,7 +2244,6 @@ void caml_domain_terminate(bool last)
   terminate_backup_thread(domain_self);
   caml_plat_unlock(&domain_self->domain_lock);
 
-  caml_plat_assert_all_locks_unlocked();
```
Comment (gasche) [inline runtime/domain.c:2316]: Can you explain why this line was removed?

File: runtime/sync.c [hunk 14]
```diff
@@ -31,23 +31,27 @@
...
-  errlen = strlen(err);
-  str = caml_alloc_string(msglen + 2 + errlen);
-  memcpy(&Byte(str, 0), msg, msglen);
-  memcpy(&Byte(str, msglen), ": ", 2);
-  memcpy(&Byte(str, msglen + 2), err, errlen);
-  caml_raise_sys_error(str);
+  char * err = caml_strerror(retcode, buf, sizeof(buf));
+  size_t msglen = strlen(msg);
+  size_t errlen = strlen(err);
+  value str = caml_alloc_string(msglen + 2 + errlen);
```
Comment (gasche) [inline runtime/sync.c:45]: This can `caml_raise_out_of_memory`, right?

**Review by gasche — COMMENTED** [review]
I have done a review pass on this PR.

Remarks:

- I now understand the broad approach -- and agree it's not so scary.
- I made peace with the use of encoded exception, I think it's okay to have them and then try to get rid of them later.
- I have not reviewed "Preparation: caml_thread_new_descriptor_exn" in details yet
- On the big picture I have remaining questions. In some places you are careful to collect exceptions to return them later (adding `_exn` functions), but in others you simply convert exception into fatal errors (for example moving from `sync_*` locks to `caml_plat_*` locks does this).
   + If we consider that having fatal errors is okay (it avoids leaks that would occur should the raised exception be caught), then I wonder if we should simplify the code further by raising them in more places -- for example do we really need to modify our external interface for `caml_domain_initialize_hook` if we could simply fatal-error from this function?
   + Or maybe the ideal situation would be to not fatal error and always raise exceptions, but then further changes would be required here and in many other places in the runtime.
   + Or maybe there are certain category of runtime failures that we consider exceedingly rare, like `pthread` failures, and we fatal-error on them, and other that we consider less rare (like `malloc` failures?) and we try to raise an exception?

   I would be curious to know what is the intended design here.

**Comment by gasche** [comment]
(Ah, something I forgot to mention: in general I am supporting of the refactorings with auxiliary functions or change to the error-code logic, they all felt in good taste to me. The removal of the `newdom` field, because it is currently unused, would conflict with #14162 where I make use of this field; but I can always revert this commit specifically within #14162 after this one gets merged.)

**Comment by gadmm** [comment]
> On the big picture I have remaining questions. In some places you are careful to collect exceptions to return them later (adding `_exn` functions), but in others you simply convert exception into fatal errors (for example moving from `sync_*` locks to `caml_plat_*` locks does this).
> 
>  * If we consider that having fatal errors is okay (it avoids leaks that would occur should the raised exception be caught), then I wonder if we should simplify the code further by raising them in more places -- for example do we really need to modify our external interface for `caml_domain_initialize_hook` if we could simply fatal-error from this function?
> 
>  * Or maybe the ideal situation would be to not fatal error and always raise exceptions, but then further changes would be required here and in many other places in the runtime.
> 
>  * Or maybe there are certain category of runtime failures that we consider exceedingly rare, like `pthread` failures, and we fatal-error on them, and other that we consider less rare (like `malloc` failures?) and we try to raise an exception?
> 
> 

I checked uses of fatal error a long time ago and there were few situations where they were used. (I have not checked recently so I do not know if the situation has changed significantly.) Bugs in the runtime are generally considered as fatal errors (e.g. using `caml_plat_*` functions for locking). We can also reasonably consider failures of small allocations as fatal, but I have seen contributors sometimes request that such allocations raise instead. (I am also not sure of what counts as small.)

Conversely, what may reasonably occur in an OCaml program, even due to programming errors of the user and to resource exhaustion of the OS, should be raised as an exception (apart from OOM situations, per design of the GC).

So we can discuss whether some errors are better considered as fatal since this can indeed simplify the code.

**Comment by gadmm** [comment]
The errors in `caml_domain_initialize_hook` do not clearly fall in either category. To give an example, `st_masterlock_init` has error codes that seem non-bug-related according to POSIX, but in the case of Linux it never fails (according to the documentation for `pthread_{mutex,cond}_init`). (It is treated as fatal in `caml_thread_reinitialize`, but this is a red herring since this is a best-effort special case documented as broken.) It is possible that someone with deeper experience of POSIX better knows which error conditions are relevant and which ones are obsolete or only theoretical, and could suggest simplifications. But I think erring on the side of caution and avoiding platform-specific assumptions is a sound approach.

**Comment by gadmm** [comment]
Let me know if you prefer that I remove the clean-up commit removing `new_dom`.

File: runtime/domain.c (hunk 11 above)
Comment (gadmm) [reply runtime/domain.c]: done, thanks for the suggestion

File: runtime/domain.c (hunk 12 above)
Comment (gadmm) [reply runtime/domain.c]: fixed!

File: runtime/domain.c (hunk 13 above)
Comment (gadmm) [reply runtime/domain.c:2316]: `terminate_domain` moves from a `caml_mutex_unlock(mut)` to a `caml_plat_unlock(mut)`, the latter of which uses `DEBUG_LOCK`. `caml_plat_assert_all_locks_unlocked` must therefore be moved right after the very last `caml_plat_unlock`. This is where `caml_plat_assert_all_locks_unlocked` is reintroduced in the subsequent commit. (Arguably removing this line would have belonged better to the subsequent commit, but we've now clarified that point.)

File: runtime/sync.c (hunk 14 above)
Comment (gadmm) [reply runtime/sync.c:45]: If `msglen` is lesser than, say, 1000, then `msglen + 2 + errlen` is lesser than `2040` which suffices to guarantee that `caml_alloc_string` allocates on the minor heap. In this case there should be no OOM exception (only fatal errors). (I have not checked that the new multicore GC indeed never raises an exception* in this case, but I remembered that people agreed that this is the desired behaviour last time this was discussed. *: that would prove fatal anyway.)

Also you are going to say that it is better if we had a noexc variant of `caml_alloc_string`, which I would agree, but I prefer to say that this PR stops there.

Comment (gasche) [reply runtime/sync.c:45]: My intuition is that this would deserve a comment, and why not an assertion / dynamic check that ensures that we are allocating in the young generation. @gadmm seems to have a slightly different opinion and I will defer to his preference.

**Comment by gasche** [comment]
I now understand @gadmm's position on error handling, which sounds reasonable. All that is left is for me the review the missing commit ("Preparation: caml_thread_new_descriptor_exn").

File: otherlibs/systhreads/st_stubs.c [hunk 15]
```diff
@@ -679,9 +689,23 @@ static void thread_init_current(caml_thread_t th)
 {
   st_tls_set(caml_thread_key, th);
   restore_runtime_state(th);
+  /* FIXME: deal with errors */
```
Comment (gasche) [inline otherlibs/systhreads/st_stubs.c]: (it may be useful to say explicitly which errors, which is not obvious to non-experts: I assume they are the ones in `caml_init_signal_stack`. @gadmm mentions that it can return `NULL` on error.)

**Review by gasche — APPROVED** [review]
I am done looking at the last commit and I believe this is okay. This completes my review of the PR. I produced a few nitpick comments and will let @gadmm decide how to best handle them.

File: otherlibs/systhreads/st_stubs.c (hunk 15 above)
Comment (gadmm) [reply otherlibs/systhreads/st_stubs.c]: done

File: runtime/sync.c (hunk 14 above)
Comment (gadmm) [reply runtime/sync.c:45]: I added a comment

**Comment by gadmm** [comment]
(I have rebased and addressed comments.)
